File: generated_code/GeminiPro/Python/068_SplayTree.py
class SplayTree:
    """
    A complete, self-contained implementation of a Splay Tree.

    This class provides a dictionary-like set for storing unique integer keys.
    It supports insertion, deletion, and searching. The key feature is the
    splaying operation: whenever a key is accessed (searched, inserted, or
    deleted), the corresponding node is moved to the root of the tree to
    optimize for future accesses.
    """

    class _Node:
        """A node in the splay tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on the given node x."""
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x
        
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        
        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on the given node x."""
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y

        y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splay operation on node x, moving it to the root.
        This is done by a series of rotations (Zig, Zig-Zig, Zig-Zag)
        until the node x is the root of the tree.
        """
        while x.parent:
            parent = x.parent
            grandparent = parent.parent
            if not grandparent:  # Zig case
                if x == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif x == parent.left and parent == grandparent.left:  # Zig-Zig (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif x == parent.right and parent == grandparent.right:  # Zig-Zig (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif x == parent.right and parent == grandparent.left:  # Zig-Zag (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-Zag (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree and performs the splaying operation.
        
        If the key is found, the node containing the key is splayed to the root.
        If the key is not found, the last non-null node visited during the search
        (the would-be parent) is splayed to the root.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        node = self.root
        last_visited = None
        while node:
            last_visited = node
            if key == node.key:
                self._splay(node)
                return True
            elif key < node.key:
                node = node.left
            else:
                node = node.right

        # If key is not found, splay the last visited node.
        if last_visited:
            self._splay(last_visited)
        
        return False

    def insert(self, key):
        """
        Inserts a key into the splay tree.

        If the key already exists, the node is splayed to the root.
        If the key is new, it is inserted as in a standard BST and then
        the new node is splayed to the root.

        Args:
            key: The integer key to insert.
        """
        node = self.root
        parent = None
        while node:
            parent = node
            if key == node.key:
                # Key already exists, splay it and return.
                self._splay(node)
                return
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        
        # Insert the new node.
        new_node = self._Node(key)
        new_node.parent = parent
        
        if not parent:
            self.root = new_node
        elif key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
            
        # Splay the newly inserted node to the root.
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the splay tree.

        The tree is first searched for the key, which brings the node (or its
        closest neighbor) to the root. If the key is present, the node (now at the
        root) is removed, and its two subtrees are joined.

        Args:
            key: The integer key to delete.
        """
        # Search for the key. This splays the node to the root if found.
        if not self.search(key):
            # Key not in the tree. search() has already splayed the closest node.
            return
        
        # After a successful search, the node to delete is at the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If there's no left subtree, the right subtree becomes the new tree.
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        else:
            # Find the maximum element in the left subtree.
            # This element will become the new root.
            left_subtree.parent = None  # Disconnect from old root
            
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Splay this max_node. It becomes the new root of the combined tree.
            # We temporarily set the tree's root to the left_subtree to perform
            # the splay within that context.
            self.root = left_subtree
            self._splay(max_node)
            
            # Now self.root is max_node. Re-attach the original right subtree.
            # After splaying the max node, its right child is guaranteed to be None.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root

File: generated_code/GeminiPro/Java/068_SplayTree.java
/**
 * A complete, self-contained implementation of a Splay Tree for an integer set.
 * Splay trees are self-balancing binary search trees with the additional property
 * that recently accessed elements are quick to access again. This implementation
 * uses a top-down splaying approach.
 */
public class SplayTree {

    private Node root;

    /**
     * Represents a node in the Splay Tree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        /**
         * Constructs a new node with the given key.
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Right-rotates the subtree rooted at y.
     * @param y The root of the subtree to rotate.
     * @return The new root of the subtree.
     */
    private Node rightRotate(Node y) {
        Node x = y.left;
        y.left = x.right;
        x.right = y;
        return x;
    }

    /**
     * Left-rotates the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * Performs the splay operation. Moves the node with the specified key,
     * or the last accessed node if the key is not found, to the root of the tree.
     * This is a top-down splay implementation.
     * @param key The key to splay around.
     */
    private void splay(int key) {
        if (root == null) {
            return;
        }

        // Dummy node to simplify linking
        Node header = new Node(0);
        Node leftTreeMax = header;
        Node rightTreeMin = header;
        Node current = root;

        while (true) {
            if (key < current.key) {
                if (current.left == null) {
                    break;
                }
                // Zig-Zig (Left-Left)
                if (key < current.left.key) {
                    current = rightRotate(current);
                    if (current.left == null) {
                        break;
                    }
                }
                // Link to the right tree
                rightTreeMin.left = current;
                rightTreeMin = current;
                current = current.left;
            } else if (key > current.key) {
                if (current.right == null) {
                    break;
                }
                // Zig-Zig (Right-Right)
                if (key > current.right.key) {
                    current = leftRotate(current);
                    if (current.right == null) {
                        break;
                    }
                }
                // Link to the left tree
                leftTreeMax.right = current;
                leftTreeMax = current;
                current = current.right;
            } else { // key == current.key
                break;
            }
        }

        // Reassemble the tree
        leftTreeMax.right = current.left;
        rightTreeMin.left = current.right;
        current.left = header.right;
        current.right = header.left;

        root = current;
    }

    /**
     * Inserts a key into the set. If the key already exists, the tree remains unchanged.
     * After insertion, the new node becomes the root of the tree.
     * @param key The key to insert.
     */
    public void insert(int key) {
        Node newNode = new Node(key);

        if (root == null) {
            root = newNode;
            return;
        }

        // Splay the tree around the key. The closest node will be at the root.
        splay(key);

        // If key is already present, do nothing (as this is a set).
        if (root.key == key) {
            return;
        }

        if (key < root.key) {
            newNode.left = root.left;
            newNode.right = root;
            root.left = null;
            root = newNode;
        } else { // key > root.key
            newNode.right = root.right;
            newNode.left = root;
            root.right = null;
            root = newNode;
        }
    }

    /**
     * Deletes a key from the set. If the key is not found, the tree remains unchanged.
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree around the key.
        splay(key);

        // If the key is not in the tree, do nothing.
        if (root.key != key) {
            return;
        }

        // Now the node to be deleted is the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
        } else {
            // Make the left subtree the new tree.
            root = leftSubtree;
            // Splay for the key we are deleting. Since this key is larger
            // than any key in the new tree (the left subtree), this will
            // bring the maximum element of the left subtree to the root.
            splay(key);
            // The new root (max of left subtree) will have a null right child.
            // Attach the original right subtree there.
            root.right = rightSubtree;
        }
    }

    /**
     * Searches for a key in the set. If the key is found, the corresponding node
     * is splayed to the root of the tree.
     * @param key The key to search for.
     * @return The key if it is found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        splay(key);
        if (root.key == key) {
            return key;
        }
        return null;
    }
}